R/cellbrowser.R

```diff
@@ -0,0 +1,499 @@
+#' parameter is passed, the function runs \code{cbBuild} (if it is
+#' installed) to create this static website in \code{cb.dir} directory.
+#' If \code{port} parameter is passed, it also runs the webserver for
+#' that directory and opens a browser.
+#'
+#' @author Maximilian Haeussler, Nikolay Markov
+#'
+#' @importFrom tools file_ext
+#' @importFrom utils browseURL
+#' @importFrom reticulate py_module_available import
+#' @importFrom Seurat Project Idents GetAssayData Embeddings FetchData
+#'
+#' @export
+#'
+#' @name CellBrowser
+#' @rdname CellBrowser
+#'
+#' @importFrom methods slot
+#' @importFrom utils packageVersion
+#' @importFrom reticulate py_module_available import
+#'
+#' @examples
+#' \dontrun{
+#' ExportToCellbrowser(pbmc_small, dataset.name = "PBMC", dir = "out")
+#' }
+#'
+ExportToCellbrowser <- function(
+  object,
+  dir,
+  dataset.name = Project(object = object),
+  reductions = NULL,
+  markers.file = NULL,
+  cluster.field = NULL,
+  cb.dir = NULL,
+  port = NULL,
+  use.mtx = FALSE,
+  meta.fields = NULL,
+  meta.fields.names = NULL,
+  matrix.slot = "counts",
+  markers.n = 100,
+  skip.markers = FALSE,
+  skip.expr.matrix = FALSE,
+  skip.metadata = FALSE,
+  skip.reductions = FALSE
+) {
+  if (!requireNamespace("Seurat", quietly = TRUE)) {
+    stop("This script requires that Seurat (V2 or V3) is installed")
+  }
+  message("Seurat Version installed: ", packageVersion(pkg = "Seurat"))
+  object.version <- as.character(x = slot(object = object, name = 'version'))
+  message("Object was created with Seurat version ", object.version)
+  if (substr(x = object.version, start = 1, stop = 1) != '2' && substr(x = object.version, start = 1, stop = 1) != '3') {
+    stop("can only process Seurat2 or Seurat3 objects, version of rds is ", object.version)
+  }
+  if (substr(x = object.version, start = 1, stop = 1) != substr(x = packageVersion(pkg = "Seurat"), start = 1, stop = 1)) {
+    stop("The installed version of Seurat is different from Seurat object loaded. You have to down- or upgrade your installed Seurat version, see the Seurat documentation")
+  }
+  embeddings <- reductions # Seurat prefers the word reductions
+  # compatibility layer for Seurat 2 vs 3
+  # see https://satijalab.org/seurat/essential_commands.html
+  if (substr(x = object.version, start = 1, stop = 1) == '2') {
+    # Seurat 2 data access
+    idents <- object@ident # Idents() in Seurat3
+    meta <- object@meta.data
+    cellOrder <- object@cell.names
+    if (matrix.slot == "counts") {
+      counts <- object@raw.data
+    } else if (matrix.slot == "scale.data") {
+      counts <- object@scale.data
+    }
+    else if (matrix.slot == "data") {
+      counts <- object@data
+    } else {
+      stop("matrix.slot can only be one of: counts, scale.data, data")
+    }
+    genes <- rownames(x = object@data)
+    dr <- object@dr
+  } else {
+    # Seurat 3 functions
+    idents <- Idents(object = object)
+    meta <- object[[]]
```

Comment:
I changed this from `object@meta.data` to match the Seurat v3 way of pulling the metadata data frame

Reply:
OK. (doesn't seem more readable to me than @meta.data, but OK)

Reply:
It is a little obscure, but we encourage the use of accessors/setters over directly accessing slots. This gives us the freedom to modify the object structure as needed while preserving data access code for the user/developer. Basically, if we decide that we need to move or restructure the cell-level metadata at the object level, we can as long as we update the `[[` method to return a data frame the way our users are used to.

For more details about getters and setters in Seurat v3, please see our [command cheatsheet](https://satijalab.org/seurat/essential_commands.html)